### PR TITLE
Exposing memory.numa_stats

### PR DESCRIPTION
**Pull request by iwankgb**
Making information on page usage by type and NUMA node available.

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
Code looks good, but needs rebase

**Comment by iwankgb** [comment]
@AkihiroSuda rebased :)

**Comment by AkihiroSuda** [comment]
LGTM

**Comment by iwankgb** [comment]
@caniszczyk, @crosbymichael, @cyphar, @dqminh, @hqhq, @mrunalp, @rjnagal, @vmarmol - second opinion would be appreciated.

File: libcontainer/cgroups/fs/memory.go [hunk 1]
```diff
@@ -269,3 +288,76 @@ func getMemoryData(path, name string) (cgroups.MemoryData, error) {
...
+
+				statsType = pagesByNode[numaStatTypeIndex]
+			}
+
+			err := addNUMAStatsByType(&stats, statsByType, statsType)
+			if err != nil {
+				return cgroups.PageUsageByNUMA{}, err
+			}
+		}
+	}
```
Comment (kolyshkin) [inline libcontainer/cgroups/fs/memory.go:336]: error checking for scanner is missing here

File: libcontainer/cgroups/fs/memory.go [hunk 2]
```diff
@@ -16,10 +19,19 @@ import (
...
-	cgroupMemoryLimit     = "memory.limit_in_bytes"
+	numaStatColumnSeparator   = " "
+	numaStatKeyValueSeparator = "="
+	numaStatMaxColumns        = math.MaxUint8 + 1
+	numaStatValueIndex        = 1
+	numaStatTypeIndex         = 0
+	numaStatColumnSliceLength = 2
+	cgroupMemorySwapLimit     = "memory.memsw.limit_in_bytes"
+	cgroupMemoryLimit         = "memory.limit_in_bytes"
+	cgroupMemoryPagesByNuma   = "memory.numa_stat"
```
Comment (kolyshkin) [inline libcontainer/cgroups/fs/memory.go:30]: Is this available in both cgroup v1 and v2?

File: libcontainer/cgroups/fs/stats_util_test.go [hunk 3]
```diff
@@ -26,98 +25,109 @@ func blkioStatEntryEquals(expected, actual []cgroups.BlkioStatEntry) error {
 
 func expectBlkioStatsEquals(t *testing.T, expected, actual cgroups.BlkioStats) {
 	if err := blkioStatEntryEquals(expected.IoServiceBytesRecursive, actual.IoServiceBytesRecursive); err != nil {
-		logrus.Printf("blkio IoServiceBytesRecursive do not match - %s\n", err)
-		t.Fail()
+		t.Errorf("blkio IoServiceBytesRecursive do not match - %s\n", err)
```
Comment (kolyshkin) [inline libcontainer/cgroups/fs/stats_util_test.go:28]: Please move this cleanup out to a separate commit that goes first.

Comment (iwankgb) [reply libcontainer/cgroups/fs/stats_util_test.go:28]: Aren't all PR's commits squashed before merge anyway?

File: libcontainer/cgroups/fs/memory.go (hunk 2 above)
Comment (kolyshkin) [reply libcontainer/cgroups/fs/memory.go:30]: looks like it's v1-only

File: libcontainer/cgroups/fs/memory.go [hunk 4]
```diff
@@ -269,3 +288,76 @@ func getMemoryData(path, name string) (cgroups.MemoryData, error) {
...
+		columns := strings.SplitN(scanner.Text(), numaStatColumnSeparator, numaStatMaxColumns)
+
+		for _, column := range columns {
+			pagesByNode := strings.SplitN(column, numaStatKeyValueSeparator, numaStatColumnSliceLength)
+			if len(pagesByNode) != numaStatColumnSliceLength {
+				return cgroups.PageUsageByNUMA{}, fmt.Errorf("wrong data format, found %d fields instead of 2", len(pagesByNode))
+			}
+
+			if isNUMANode.MatchString(pagesByNode[numaStatTypeIndex]) {
+				nodeID, err := strconv.ParseUint(pagesByNode[numaStatTypeIndex][1:], 10, 8)
```
Comment (kolyshkin) [inline libcontainer/cgroups/fs/memory.go:313]: I think this check can be simplified by just checking for capital `N` plus no error from `ParseUint`.

File: libcontainer/cgroups/fs/memory.go [hunk 5]
```diff
@@ -269,3 +288,76 @@ func getMemoryData(path, name string) (cgroups.MemoryData, error) {
...
+	}
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		var statsType string
+		statsByType := cgroups.PageStats{Nodes: map[uint8]uint64{}}
+		columns := strings.SplitN(scanner.Text(), numaStatColumnSeparator, numaStatMaxColumns)
+
+		for _, column := range columns {
+			pagesByNode := strings.SplitN(column, numaStatKeyValueSeparator, numaStatColumnSliceLength)
```
Comment (kolyshkin) [inline libcontainer/cgroups/fs/memory.go:307]: since `numaStatKeyValueSeparator` is only used once, it's better to inline it, e.g.
```go
strings.SplitN(column, "=", 2)
```

File: libcontainer/cgroups/fs/memory.go [hunk 6]
```diff
@@ -269,3 +288,76 @@ func getMemoryData(path, name string) (cgroups.MemoryData, error) {
...
+			if len(pagesByNode) != numaStatColumnSliceLength {
```
Comment (kolyshkin) [inline libcontainer/cgroups/fs/memory.go:308]: If you have `column` as `A=B=C`, then `len(pagesByNode)` will still be 2 since you've used SplitN with the value of 2.

File: libcontainer/cgroups/fs/memory.go (hunk 2 above)
Comment (iwankgb) [reply libcontainer/cgroups/fs/memory.go:30]: I can't see it in v2.

**Comment by iwankgb** [comment]
@AkihiroSuda I would appreciate one more iteration of review.

File: libcontainer/cgroups/fs/memory.go (hunk 5 above)
Comment (iwankgb) [reply libcontainer/cgroups/fs/memory.go:305]: All other constants are used once, I think. I still prefer them than magic strings.

**Comment by iwankgb** [comment]
Travis build succeeded but status has not been reported: https://travis-ci.org/github/opencontainers/runc/builds/672578771.

**Comment by AkihiroSuda** [comment]
LGTM

**Comment by crosbymichael** [comment]
LGTM
